OrbitControls: Cancel user interaction on disable

gkjohnson opened the pull request:
My application currently disables orbit controls when the camera needs to be taken control of or the user is interacting with something else in the scene. Currently, however, when disabling OrbitControls they are left in the last drag state and never updated until they are reenabled again meaning it can miss pointer events like "mouseup" that would reset the state back to "NONE". In those cases if the controls must be disabled after the mousedown event is fired it will be in the drag state when they are reenabled and could result in the camera orbiting while the mouse is not clicked.

The `reset()` function _kind of_ addresses this but also resets the camera position and does not remove lingering event listeners.

This PR adds a `cancel` function to cancel the current user interaction if one is happening and clean up the event listeners the way it does in the `mouseup` callback. Just using `enabled = false` will still result in the behavior mentioned above unless `cancel` is called.

I tested this by adding a set interval every two seconds to call controls.cancel in the orbit controls example with damping both enabled and disabled.

gkjohnson committed:
Add cancel function to orbit controls

gkjohnson committed:
modularize

Mugen87 commented:
Does this change also works with touch devices? It seems the problem you have described could also occur with touch interaction.

gkjohnson committed:
call cancel in touchend

gkjohnson commented:
> Does this change also works with touch devices? It seems the problem you have described could also occur with touch interaction.

Oops didn't realize touch was using separate events in OrbitControls -- the current change would have still worked but I've modified `onTouchEnd` to call `cancel` for consistency. Touch events are not dynamically added and removed like they are for pointer events so there was nothing extra that had to be added to the cancel implementation. I tested touch the same way as outlined above but with the "mobile" mode in chrome.

Thanks!

Mugen87 commented on examples/js/controls/OrbitControls.js:
```diff
@@ -116,7 +116,22 @@ THREE.OrbitControls = function ( object, domElement ) {
 		scope.update();
 
-		state = STATE.NONE;
+		scope.cancel();
```
I'm not sure it makes sense to fire an `end` event in `reset()`. How about keeping `reset()` as it was and only focus on the event listeners?

gkjohnson replied:
My feeling was that for every "start" event we should be guaranteed an "end" event when the orbit controls interaction stops, which is why I added an "end" dispatch only when the controls are in the middle of a drag. Without this (and as can happen now) it's possible to get two "start" events dispatched back to back.

gkjohnson replied:
@Mugen87 the structure of the fix has changed a bit -- do you have any further thoughts on how this should be handled?

WestLangley commented:
I think the user should only be canceling event listeners created by the user.

Perhaps implement `disable` via a setter, and automatically handle it there.

arodic commented:
> Perhaps implement disable via a setter, and automatically handle it there.

That is a solution that worked great for my use case.

gkjohnson commented:
@WestLangley 
> I think the user should only be canceling event listeners created by the user.

Can you elaborate? In order to properly fully cancel the active drag interaction the events must be removed as they are when "mouse up" is called. If the drag state is set to `None` without removing the events as happens in `reset` in master then you'll technically have lingering events left registered that aren't normally there. And then when the user starts dragging again [two more events will be registered](https://github.com/mrdoob/three.js/blob/e08066c97ae4f0c3425cc03ba101293eae1b9ae3/examples/js/controls/OrbitControls.js#L934-L935) meaning there will be two copies of `pointerup` and `pointermove` events registered until they're removed on mouse up which doesn't seem right.

> Perhaps implement disable via a setter, and automatically handle it there.

Do you mean doing something like this to recreate the cancel code when `controls.enable` is set to false?

```js
Object.defineProperty(
  scope,
  'enabled',
  {
    get() { return scope._enabled },
    set( v ) {
      scope._enabled = v;
      if ( ! v ) {
        // stop any active interaction
      }
    }
  }
);
```

@arodic 
> That is a solution that worked great for my use case.

Can you link an example of what you're doing?

Thanks for the feedback!

arodic commented:
> Can you link an example of what you're doing?

Sure, [here](https://github.com/io-gui/three.js/blob/pointerEvents/orbitControls/examples/jsm/controls/Controls.js#L50) I define `enabled` property with callback to `_connect()` and `_disconnect()` on property change. Bun unlike most threejs controls, I dont add listeners on pointerdown, instead I use PointerCapture API which is designed to simplify pointer capturing.

WestLangley commented:
@gkjohnson What @arodic implemented is what I was thinking of. Thank you, @arodic for sharing your example. :-)

gkjohnson committed:
disable movement on enabled toggle

gkjohnson committed:
set enabled to the set value

gkjohnson committed:
make controls events resilient to enabled toggle mid drag

gkjohnson commented:
Okay I've simplified the change to behave in the way you requested and updated the touch up and mouse up callbacks to guard against dispatching end events in the case that the state is set to `NONE`.

I tested by exposing controls in the `misc_controls_orbit.html` example and running the following in the console to ensure the that drags are stopped every 2 seconds:

```js
controls.addEventListener('start', () => console.log('start'));
controls.addEventListener('end', () => console.log('end'));
setInterval(() => {
  controls.enabled = false;
  controls.enabled = true;
}, 2000)
```

Let me know what you think!

gkjohnson commented:
@WestLangley just wanted to ping again to see if the the changes are more along the lines of what you were thinking. Thanks!

WestLangley commented:
@gkjohnson Yes. Thank you for trying a different approach. I will leave final code approval to others.

gkjohnson commented:
ping @Mugen87 @mrdoob -- any further feedback on this PR? Thanks!

mrdoob commented:
#20380 solved this.
